### PR TITLE
MDEV-9903 : Check and run rsync daemon only in the needed IP

### PR DESCRIPTION
**Pull request by fmbiete**
I submit this code under the BSD-new license.

### CODE REVIEW

**Comment by svoj** [comment]
Hi Francisco,

Thanks for your contribution. JIRA task just for the record: https://jira.mariadb.org/browse/MDEV-9903

This task was added to 10.1.19 backlog, which is planned to be handled between 2016-10-13 and 2016-10-20.

Thanks,
Sergey

**Comment by michaeldg** [comment]
Hey Francisco,

I don't think your contribution is compatible with all set-ups: Imagine that it's a WAN set-up running behind a NAT firewall. The wsrep_sst_receive_address must be the global IP but that IP is not bound to the server where MariaDB is running.

I do think this change is useful for security. The rsync is not secured, if a customer does not run a firewal on this port an intruder could, if timed correctly, catch the rsync stream. I think it should be enabled by default but configurable.

Can you change this so it supports enabling/disabling this in the config file? There is some SST configuration in the xtrabackup-v2 SST scripts already, for example to enable throttle or change the network transfer method.

Cheers,
Michaël

**Comment by fmbiete** [comment]
Hi Michael,

We could check if the sst address for the donor is local (ip or ifconfig). If it is, start rsync in that interface. If it's not start rsync in all.
That way it would be automatic and don't need any configuration option.
Would that be ok?

I don't understand the bit about security. This was never intended to provide security, but allow to execute rsync sst donor on a server that already has a rsync server for other stuff.

Regards,

File: scripts/wsrep_sst_rsync.sh [hunk 1]
```diff
@@ -69,18 +70,42 @@ check_pid_and_port()
...
     fi
     check_pid $pid_file && \
         [ -n "$port_info" ] && [ -n "$is_rsync" ] && \
         [ $(cat $pid_file) -eq $rsync_pid ]
 }
 
+is_local_ip()
+{
+  local address="$1"
+  local wherebin=`which ifconfig`
```
Comment (nirbhayc) [inline scripts/wsrep_sst_rsync.sh]: wherebin sounds a bit vague. Perhaps get_addr would be more appropriate?

File: scripts/wsrep_sst_rsync.sh [hunk 2]
```diff
@@ -69,18 +70,42 @@ check_pid_and_port()
...
+  if [ -z "$wherebin" ]
+  then
+    wherebin=`which ip`
+    wherebin="$wherebin address show"
+    # Add an slash at the end, so we don't get false positive : 172.18.0.4 matches 172.18.0.41
+    address="${address}/"
+  else
+    # Add an space at the end, so we don't get false positive : 172.18.0.4 matches 172.18.0.41
+    address="$address "
```
Comment (nirbhayc) [inline scripts/wsrep_sst_rsync.sh]: Could you also add a comment to explain why do we need 2 different address terminators?

**Comment by fmbiete** [comment]
I have created merge request #241 against branch 10.2, with the requested changes.
This one can be closed then.

**Pull request closed without merging**